File: frameworks/android/src/main/java/com/einstore/tracking/EinstoreTracker.kt
package com.einstore.tracking

import android.content.Context
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.os.Build
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.util.Locale
import java.util.UUID
import java.util.concurrent.Executor
import java.util.concurrent.Executors

enum class EinstoreService(val key: String) {
  ANALYTICS("analytics"),
  ERRORS("errors"),
  DISTRIBUTION("distribution"),
  DEVICES("devices"),
  USAGE("usage"),
  CRASHES("crashes");

  companion object {
    fun defaults(): Set<EinstoreService> = setOf(DISTRIBUTION, DEVICES)
  }
}

data class EinstoreTrackingConfig(
  val baseUrl: String? = "https://api.einstore.dev",
  val buildId: String? = null,
  val downloadUrl: String? = null,
  val launchUrl: String? = null,
  val eventUrl: String? = null,
  val headers: Map<String, String> = emptyMap(),
  val platform: String = "android",
  val targetId: String? = null,
  val deviceId: String? = null,
  val metadata: Map<String, Any?> = emptyMap(),
  val services: Set<EinstoreService> = EinstoreService.defaults(),
  val distributionInfo: Map<String, Any?> = emptyMap(),
  val deviceInfo: Map<String, Any?> = emptyMap(),
  val launchKey: String? = null,
  val crashEnabled: Boolean = false
)

class EinstoreTracker(
  context: Context,
  private val config: EinstoreTrackingConfig,
  private val executor: Executor = Executors.newSingleThreadExecutor()
) {
  private data class AnalyticsEvent(val name: String, val properties: Map<String, Any?>)
  private data class ErrorInfo(val message: String, val stackTrace: String?, val properties: Map<String, Any?>)

  private val appContext = context.applicationContext
  private val preferences = appContext.getSharedPreferences("einstore_tracking", Context.MODE_PRIVATE)
  private val crashKey = "einstore_crashes"
  private var sessionId = UUID.randomUUID().toString()
  private var sessionStartMs = System.currentTimeMillis()
  private val userProperties = mutableMapOf<String, Any?>()

  init {
    if (config.crashEnabled && config.services.contains(EinstoreService.CRASHES)) {
      uploadPendingCrashes()
    }
  }

  fun startNewSession() {
    sessionId = UUID.randomUUID().toString()
    sessionStartMs = System.currentTimeMillis()
  }

  fun setUserProperties(properties: Map<String, Any?>) {
    userProperties.putAll(properties)
  }

  fun trackDownload(callback: ((Result<Unit>) -> Unit)? = null) {
    trackAsync(resolvedDownloadUrl(), requiredService = EinstoreService.DISTRIBUTION, callback = callback)
  }

  fun trackLaunch(callback: ((Result<Unit>) -> Unit)? = null) {
    startNewSession()
    trackAsync(
      resolvedLaunchUrl(),
      event = AnalyticsEvent("app_launch", emptyMap()),
      requiredService = EinstoreService.ANALYTICS,
      callback = callback
    )
  }

  fun trackLaunchOnce(callback: ((Result<Unit>) -> Unit)? = null) {
    val key = resolveLaunchKey()
    if (preferences.getBoolean(key, false)) {
      callback?.invoke(Result.success(Unit))
      return
    }
    trackLaunch { result ->
      if (result.isSuccess) {
        preferences.edit().putBoolean(key, true).apply()
      }
      callback?.invoke(result)
    }
  }

  fun trackScreenView(
    screenName: String,
    properties: Map<String, Any?> = emptyMap(),
    callback: ((Result<Unit>) -> Unit)? = null
  ) {
    val payload = properties.toMutableMap()
    payload["screen"] = screenName
    trackAsync(
      resolvedEventUrl(),
      event = AnalyticsEvent("screen_view", payload),
      requiredService = EinstoreService.ANALYTICS,
      callback = callback
    )
  }

  fun trackEvent(
    name: String,
    properties: Map<String, Any?> = emptyMap(),
    callback: ((Result<Unit>) -> Unit)? = null
  ) {
    trackAsync(
      resolvedEventUrl(),
      event = AnalyticsEvent(name, properties),
      requiredService = EinstoreService.ANALYTICS,
      callback = callback
    )
  }

  fun trackError(
    message: String,
    stackTrace: String? = null,
    properties: Map<String, Any?> = emptyMap(),
    callback: ((Result<Unit>) -> Unit)? = null
  ) {
    trackAsync(
      resolvedEventUrl(),
      errorInfo = ErrorInfo(message, stackTrace, properties),
      requiredService = EinstoreService.ERRORS,
      callback = callback
    )
  }

  fun recordCrash(crash: Map<String, Any?>) {
    if (!config.crashEnabled || !config.services.contains(EinstoreService.CRASHES)) return
    val existing = pendingCrashes().toMutableList()
    existing.add(crash)
    preferences.edit().putString(crashKey, JSONArray(existing).toString()).apply()
  }

  fun uploadPendingCrashes(callback: ((Result<Unit>) -> Unit)? = null) {
    if (!config.crashEnabled || !config.services.contains(EinstoreService.CRASHES)) {
      callback?.invoke(Result.success(Unit))
      return
    }
    val crashes = pendingCrashes()
    if (crashes.isEmpty()) {
      callback?.invoke(Result.success(Unit))
      return
    }
    val results = mutableListOf<Result<Unit>>()
    crashes.forEach { crash ->
      trackAsync(resolvedEventUrl(), crash = crash, requiredService = EinstoreService.CRASHES) {
        results.add(it)
        if (results.size == crashes.size) {
          if (results.any { res -> res.isFailure }) {
            callback?.invoke(results.first { it.isFailure })
          } else {
            preferences.edit().remove(crashKey).apply()
            callback?.invoke(Result.success(Unit))
          }
        }
      }
    }
  }

  private fun resolvedDownloadUrl(): String? {
    return resolveUrl(config.downloadUrl, "builds/{id}/downloads")
  }

  private fun resolvedLaunchUrl(): String? {
    return resolveUrl(config.launchUrl, "builds/{id}/installs")
  }

  private fun resolvedEventUrl(): String? {
    return resolveUrl(config.eventUrl, "builds/{id}/events") ?: resolveUrl(config.launchUrl, "builds/{id}/installs")
  }

  private fun resolveUrl(provided: String?, defaultPath: String): String? {
    if (!provided.isNullOrBlank()) {
      return provided
    }
    val base = config.baseUrl
    if (base.isNullOrBlank()) {
      return null
    }
    val buildId = config.buildId
    val path = if (!buildId.isNullOrBlank()) {
      defaultPath.replace("{id}", buildId)
    } else {
      "tracking/events"
    }
    return if (base.endsWith("/")) base + path else "$base/$path"
  }

  private fun trackAsync(
    url: String?,
    event: AnalyticsEvent? = null,
    errorInfo: ErrorInfo? = null,
    crash: Map<String, Any?>? = null,
    requiredService: EinstoreService? = null,
    callback: ((Result<Unit>) -> Unit)?
  ) {
    if (url.isNullOrBlank()) {
      callback?.invoke(Result.failure(IllegalStateException("Tracking URL is required")))
      return
    }
    if (requiredService != null && !config.services.contains(requiredService)) {
      callback?.invoke(Result.failure(IllegalStateException("Service ${requiredService.key} is disabled")))
      return
    }
    executor.execute {
      callback?.invoke(post(url, event, errorInfo, crash))
    }
  }

  private fun post(url: String, event: AnalyticsEvent?, errorInfo: ErrorInfo?, crash: Map<String, Any?>?): Result<Unit> {
    return try {
      val payload = buildPayload(event, errorInfo, crash)
      val connection = URL(url).openConnection() as HttpURLConnection
      connection.requestMethod = "POST"
      connection.doOutput = true
      connection.setRequestProperty("Content-Type", "application/json")
      for ((key, value) in config.headers) {
        connection.setRequestProperty(key, value)
      }
      connection.outputStream.use { output ->
        output.write(payload.toByteArray(Charsets.UTF_8))
      }
      val status = connection.responseCode
      val stream = if (status in 200..299) connection.inputStream else connection.errorStream
      stream?.use { it.readBytes() }
      if (status in 200..299) {
        Result.success(Unit)
      } else {
        Result.failure(IllegalStateException("Tracking failed with status $status"))
      }
    } catch (ex: Exception) {
      Result.failure(ex)
    }
  }

  private fun buildPayload(event: AnalyticsEvent?, errorInfo: ErrorInfo?, crash: Map<String, Any?>?): String {
    val root = JSONObject()
    root.put("platform", config.platform)
    resolveDeviceId()?.let { root.put("deviceId", it) }
    resolveTargetId()?.let { root.put("targetId", it) }

    val metadata = JSONObject()
    metadata.put("services", JSONArray(config.services.map { it.key }))

    if (config.services.contains(EinstoreService.ANALYTICS)) {
      val analytics = JSONObject()
      if (event != null) {
        analytics.put("event", JSONObject().apply {
          put("name", event.name)
          if (event.properties.isNotEmpty()) {
            put("properties", mapToJson(event.properties))
          }
        })
      }
      if (userProperties.isNotEmpty()) {
        analytics.put("userProperties", mapToJson(userProperties))
      }
      analytics.put("session", sessionPayload())
      if (analytics.length() > 0) {
        metadata.put("analytics", analytics)
      }
    }

    if (config.services.contains(EinstoreService.ERRORS) && errorInfo != null) {
      val errorPayload = JSONObject().apply {
        put("message", errorInfo.message)
        if (!errorInfo.stackTrace.isNullOrBlank()) {
          put("stackTrace", errorInfo.stackTrace)
        }
        if (errorInfo.properties.isNotEmpty()) {
          put("properties", mapToJson(errorInfo.properties))
        }
      }
      metadata.put("errors", errorPayload)
    }

    if (config.services.contains(EinstoreService.DISTRIBUTION)) {
      val distribution = distributionPayload()
      if (distribution.length() > 0) {
        metadata.put("distribution", distribution)
      }
    }

    if (config.services.contains(EinstoreService.DEVICES)) {
      val device = devicePayload()
      if (device.length() > 0) {
        metadata.put("device", device)
      }
    }

    if (config.services.contains(EinstoreService.USAGE)) {
      metadata.put("usage", usagePayload())
    }

    if (config.services.contains(EinstoreService.CRASHES) && crash != null) {
      val crashPayload = mapToJson(crash.toMutableMap().apply {
        putIfAbsent("platform", config.platform)
        if (!containsKey("appVersion")) {
          put("appVersion", resolvePackageInfo()?.versionName)
        }
        if (!containsKey("buildNumber")) {
          put("buildNumber", resolveVersionCode(resolvePackageInfo()))
        }
      })
      metadata.put("crash", crashPayload)
    }

    if (config.metadata.isNotEmpty()) {
      metadata.put("custom", mapToJson(config.metadata))
    }

    if (metadata.length() > 0) {
      root.put("metadata", metadata)
    }

    return root.toString()
  }

  private fun sessionPayload(): JSONObject {
    val durationMs = System.currentTimeMillis() - sessionStartMs
    return JSONObject().apply {
      put("id", sessionId)
      put("startedAt", isoTimestamp(sessionStartMs))
      put("durationMs", durationMs)
    }
  }

  private fun usagePayload(): JSONObject {
    val now = System.currentTimeMillis()
    val offsetMinutes = (java.util.TimeZone.getDefault().getOffset(now) / 60000)
    return JSONObject().apply {
      put("timestamp", isoTimestamp(now))
      put("timeZone", java.util.TimeZone.getDefault().id)
      put("timeZoneOffsetMinutes", offsetMinutes)
      put("locale", Locale.getDefault().toLanguageTag())
      put("sessionId", sessionId)
      put("sessionDurationMs", now - sessionStartMs)
    }
  }

  private fun distributionPayload(): JSONObject {
    val payload = mutableMapOf<String, Any?>()
    payload.putAll(config.distributionInfo)
    val packageInfo = resolvePackageInfo()
    if (payload["appVersion"] == null) {
      payload["appVersion"] = packageInfo?.versionName
    }
    if (payload["buildNumber"] == null) {
      payload["buildNumber"] = resolveVersionCode(packageInfo)
    }
    return mapToJson(payload)
  }

  private fun devicePayload(): JSONObject {
    val payload = mutableMapOf<String, Any?>()
    payload.putAll(config.deviceInfo)
    if (payload["model"] == null) {
      payload["model"] = Build.MODEL
    }
    if (payload["manufacturer"] == null) {
      payload["manufacturer"] = Build.MANUFACTURER
    }
    if (payload["osVersion"] == null) {
      payload["osVersion"] = Build.VERSION.RELEASE
    }
    if (payload["locale"] == null) {
      payload["locale"] = Locale.getDefault().toLanguageTag()
    }
    val packageInfo = resolvePackageInfo()
    if (payload["appVersion"] == null) {
      payload["appVersion"] = packageInfo?.versionName
    }
    if (payload["buildNumber"] == null) {
      payload["buildNumber"] = resolveVersionCode(packageInfo)
    }
    return mapToJson(payload)
  }

  private fun resolveDeviceId(): String? {
    if (!config.deviceId.isNullOrBlank()) {
      return config.deviceId
    }
    val existing = preferences.getString("device_id", null)
    if (!existing.isNullOrBlank()) {
      return existing
    }
    val created = UUID.randomUUID().toString()
    preferences.edit().putString("device_id", created).apply()
    return created
  }

  private fun pendingCrashes(): List<Map<String, Any?>> {
    val json = preferences.getString(crashKey, null) ?: return emptyList()
    return try {
      val array = JSONArray(json)
      (0 until array.length()).mapNotNull { idx ->
        val obj = array.optJSONObject(idx) ?: return@mapNotNull null
        obj.toMap()
      }
    } catch (_: Exception) {
      emptyList()
    }
  }

  private fun resolveTargetId(): String? {
    if (!config.targetId.isNullOrBlank()) {
      return config.targetId
    }
    return appContext.packageName
  }

  private fun resolveLaunchKey(): String {
    if (!config.launchKey.isNullOrBlank()) {
      return config.launchKey
    }
    val packageName = appContext.packageName
    val packageInfo = resolvePackageInfo()
    val versionName = packageInfo?.versionName ?: "0"
    val versionCode = resolveVersionCode(packageInfo) ?: "0"
    return "einstore.launch.$packageName.$versionName.$versionCode"
  }

  private fun resolvePackageInfo(): PackageInfo? {
    return try {
      appContext.packageManager.getPackageInfo(appContext.packageName, 0)
    } catch (_: PackageManager.NameNotFoundException) {
      null
    }
  }

  private fun resolveVersionCode(packageInfo: PackageInfo?): String? {
    packageInfo ?: return null
    return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
      packageInfo.longVersionCode.toString()
    } else {
      @Suppress("DEPRECATION")
      packageInfo.versionCode.toString()
    }
  }

  private fun JSONObject.toMap(): Map<String, Any?> {
    val map = mutableMapOf<String, Any?>()
    val keys = keys()
    while (keys.hasNext()) {
      val key = keys.next()
      map[key] = this.opt(key)
    }
    return map
  }

  private fun mapToJson(map: Map<String, Any?>): JSONObject {
    val json = JSONObject()
    for ((key, value) in map) {
      if (value == null) {
        continue
      }
      when (value) {
        is Map<*, *> -> {
          @Suppress("UNCHECKED_CAST")
          json.put(key, mapToJson(value as Map<String, Any?>))
        }
        is List<*> -> {
          json.put(key, JSONArray(value))
        }
        else -> json.put(key, value)
      }
    }
    return json
  }

  private fun isoTimestamp(value: Long): String {
    val formatter = java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US)
    formatter.timeZone = java.util.TimeZone.getTimeZone("UTC")
    return formatter.format(java.util.Date(value))
  }
}

File: frameworks/ios/EinstoreTracking/Sources/EinstoreTracking/EinstoreTracker.swift
import Foundation
#if canImport(UIKit)
import UIKit
#endif

public enum EinstoreTrackingError: Error {
  case invalidPayload
  case invalidResponse
  case requestFailed(statusCode: Int)
  case missingUrl
  case serviceDisabled(String)
}

public enum EinstoreService: String, CaseIterable {
  case analytics
  case errors
  case distribution
  case devices
  case usage
  case crashes
}

public protocol EinstoreLaunchStorage {
  func hasLaunched(key: String) -> Bool
  func markLaunched(key: String)
}

public final class UserDefaultsLaunchStorage: EinstoreLaunchStorage {
  private let defaults: UserDefaults

  public init(defaults: UserDefaults = .standard) {
    self.defaults = defaults
  }

  public func hasLaunched(key: String) -> Bool {
    defaults.bool(forKey: key)
  }

  public func markLaunched(key: String) {
    defaults.set(true, forKey: key)
  }
}

public protocol EinstoreDeviceIdProvider {
  func deviceId() -> String
}

public final class UserDefaultsDeviceIdProvider: EinstoreDeviceIdProvider {
  private let key = "einstore.device-id"
  private let defaults: UserDefaults

  public init(defaults: UserDefaults = .standard) {
    self.defaults = defaults
  }

  public func deviceId() -> String {
    if let existing = defaults.string(forKey: key), !existing.isEmpty {
      return existing
    }
    let created = UUID().uuidString
    defaults.set(created, forKey: key)
    return created
  }
}

public struct EinstoreTrackingConfig {
  public let baseUrl: URL?
  public let buildId: String?
  public let downloadUrl: URL?
  public let launchUrl: URL?
  public let eventUrl: URL?
  public let headers: [String: String]
  public let platform: String
  public let targetId: String?
  public let deviceId: String?
  public let metadata: [String: Any]?
  public let launchKey: String?
  public let services: [EinstoreService]
  public let distributionInfo: [String: Any]
  public let deviceInfo: [String: Any]
  public let crashEnabled: Bool

  public init(
    baseUrl: URL? = URL(string: "https://api.einstore.dev"),
    buildId: String? = nil,
    downloadUrl: URL? = nil,
    launchUrl: URL? = nil,
    eventUrl: URL? = nil,
    headers: [String: String] = [:],
    platform: String = "ios",
    targetId: String? = nil,
    deviceId: String? = nil,
    metadata: [String: Any]? = nil,
    launchKey: String? = nil,
    services: [EinstoreService] = [.distribution, .devices],
    distributionInfo: [String: Any] = [:],
    deviceInfo: [String: Any] = [:],
    crashEnabled: Bool = false
  ) {
    self.baseUrl = baseUrl
    self.buildId = buildId
    self.downloadUrl = downloadUrl
    self.launchUrl = launchUrl
    self.eventUrl = eventUrl
    self.headers = headers
    self.platform = platform
    self.targetId = targetId
    self.deviceId = deviceId
    self.metadata = metadata
    self.launchKey = launchKey
    self.services = services
    self.distributionInfo = distributionInfo
    self.deviceInfo = deviceInfo
    self.crashEnabled = crashEnabled
  }
}

public final class EinstoreTracker {
  private struct AnalyticsEvent {
    let name: String
    let properties: [String: Any]
  }

  private struct ErrorInfo {
    let message: String
    let stackTrace: String?
    let properties: [String: Any]
  }

  private let config: EinstoreTrackingConfig
  private let session: URLSession
  private let deviceIdProvider: EinstoreDeviceIdProvider
  private let launchStorage: EinstoreLaunchStorage
  private let crashStorageKey = "einstore.crashes"
  private var sessionId: String
  private var sessionStart: Date
  private var userProperties: [String: Any]

  public init(
    config: EinstoreTrackingConfig,
    session: URLSession = .shared,
    deviceIdProvider: EinstoreDeviceIdProvider = UserDefaultsDeviceIdProvider(),
    launchStorage: EinstoreLaunchStorage = UserDefaultsLaunchStorage()
  ) {
    self.config = config
    self.session = session
    self.deviceIdProvider = deviceIdProvider
    self.launchStorage = launchStorage
    self.sessionId = UUID().uuidString
    self.sessionStart = Date()
    self.userProperties = [:]
    if config.crashEnabled && isServiceEnabled(.crashes) {
      uploadPendingCrashes()
    }
  }

  public func startNewSession() {
    sessionId = UUID().uuidString
    sessionStart = Date()
  }

  public func setUserProperties(_ properties: [String: Any]) {
    for (key, value) in properties {
      userProperties[key] = value
    }
  }

  public func trackDownload(completion: ((Result<Void, Error>) -> Void)? = nil) {
    track(url: resolvedDownloadUrl, requiredService: .distribution, completion: completion)
  }

  public func trackLaunch(completion: ((Result<Void, Error>) -> Void)? = nil) {
    startNewSession()
    track(
      url: resolvedLaunchUrl,
      event: AnalyticsEvent(name: "app_launch", properties: [:]),
      requiredService: .analytics,
      completion: completion
    )
  }

  public func recordCrash(_ crash: [String: Any]) {
    guard config.crashEnabled, isServiceEnabled(.crashes) else { return }
    var existing = pendingCrashes()
    existing.append(crash)
    UserDefaults.standard.set(existing, forKey: crashStorageKey)
  }

  public func uploadPendingCrashes(completion: ((Result<Void, Error>) -> Void)? = nil) {
    guard config.crashEnabled, isServiceEnabled(.crashes) else {
      completion?(.success(()))
      return
    }
    let crashes = pendingCrashes()
    guard !crashes.isEmpty else {
      completion?(.success(()))
      return
    }
    let group = DispatchGroup()
    var firstError: Error?
    for crash in crashes {
      group.enter()
      trackCrash(crash: crash) { result in
        if case .failure(let error) = result, firstError == nil {
          firstError = error
        }
        group.leave()
      }
    }
    group.notify(queue: .main) {
      if firstError == nil {
        UserDefaults.standard.removeObject(forKey: self.crashStorageKey)
        completion?(.success(()))
      } else if let error = firstError {
        completion?(.failure(error))
      }
    }
  }

  public func trackLaunchOnce(completion: ((Result<Void, Error>) -> Void)? = nil) {
    let key = resolveLaunchKey()
    if launchStorage.hasLaunched(key: key) {
      completion?(.success(()))
      return
    }

    trackLaunch { [launchStorage] result in
      if case .success = result {
        launchStorage.markLaunched(key: key)
      }
      completion?(result)
    }
  }

  public func trackScreenView(
    screenName: String,
    properties: [String: Any] = [:],
    completion: ((Result<Void, Error>) -> Void)? = nil
  ) {
    var payload = properties
    payload["screen"] = screenName
    track(
      url: resolvedEventUrl ?? resolvedLaunchUrl,
      event: AnalyticsEvent(name: "screen_view", properties: payload),
      requiredService: .analytics,
      completion: completion
    )
  }

  public func trackEvent(
    name: String,
    properties: [String: Any] = [:],
    completion: ((Result<Void, Error>) -> Void)? = nil
  ) {
    track(
      url: resolvedEventUrl ?? resolvedLaunchUrl,
      event: AnalyticsEvent(name: name, properties: properties),
      requiredService: .analytics,
      completion: completion
    )
  }

  public func trackError(
    message: String,
    stackTrace: String? = nil,
    properties: [String: Any] = [:],
    completion: ((Result<Void, Error>) -> Void)? = nil
  ) {
    track(
      url: resolvedEventUrl ?? resolvedLaunchUrl,
      errorInfo: ErrorInfo(message: message, stackTrace: stackTrace, properties: properties),
      requiredService: .errors,
      completion: completion
    )
  }

  private func resolveLaunchKey() -> String {
    if let launchKey = config.launchKey, !launchKey.isEmpty {
      return launchKey
    }
    let bundleId = Bundle.main.bundleIdentifier ?? "unknown"
    let buildNumber = Bundle.main.infoDictionary?["CFBundleVersion"] as? String ?? "0"
    return "einstore.launch.\(bundleId).\(buildNumber)"
  }

  private func resolveUrl(_ provided: URL?, defaultPath: String) -> URL? {
    if let provided {
      return provided
    }
    guard let base = config.baseUrl else {
      return nil
    }
    if let buildId = config.buildId {
      return base.appendingPathComponent(defaultPath.replacingOccurrences(of: "{id}", with: buildId))
    }
    return base.appendingPathComponent("tracking/events")
  }

  private var resolvedDownloadUrl: URL? {
    resolveUrl(config.downloadUrl, defaultPath: "builds/{id}/downloads")
  }

  private var resolvedLaunchUrl: URL? {
    resolveUrl(config.launchUrl, defaultPath: "builds/{id}/installs")
  }

  private var resolvedEventUrl: URL? {
    resolveUrl(config.eventUrl, defaultPath: "builds/{id}/events")
  }

  private func resolveDeviceId() -> String? {
    if let deviceId = config.deviceId, !deviceId.isEmpty {
      return deviceId
    }
    return deviceIdProvider.deviceId()
  }

  private func resolveTargetId() -> String? {
    if let targetId = config.targetId, !targetId.isEmpty {
      return targetId
    }
    return Bundle.main.bundleIdentifier
  }

  private func pendingCrashes() -> [[String: Any]] {
    (UserDefaults.standard.array(forKey: crashStorageKey) as? [[String: Any]]) ?? []
  }

  private func trackCrash(
    crash: [String: Any],
    completion: ((Result<Void, Error>) -> Void)?
  ) {
    track(
      url: resolvedEventUrl ?? resolvedLaunchUrl,
      crash: crash,
      requiredService: .crashes,
      completion: completion
    )
  }

  private func track(
    url: URL?,
    event: AnalyticsEvent? = nil,
    errorInfo: ErrorInfo? = nil,
    crash: [String: Any]? = nil,
    requiredService: EinstoreService? = nil,
    completion: ((Result<Void, Error>) -> Void)?
  ) {
    guard let url else {
      completion?(.failure(EinstoreTrackingError.missingUrl))
      return
    }
    if let requiredService, !isServiceEnabled(requiredService) {
      completion?(.failure(EinstoreTrackingError.serviceDisabled(requiredService.rawValue)))
      return
    }
    do {
      let payload = try buildPayload(event: event, errorInfo: errorInfo, crash: crash)
      send(url: url, payload: payload, completion: completion)
    } catch {
      completion?(.failure(error))
    }
  }

  private func buildPayload(event: AnalyticsEvent?, errorInfo: ErrorInfo?, crash: [String: Any]?) throws -> [String: Any] {
    var payload: [String: Any] = ["platform": config.platform]
    if let deviceId = resolveDeviceId() {
      payload["deviceId"] = deviceId
    }
    if let targetId = resolveTargetId() {
      payload["targetId"] = targetId
    }

    var metadata: [String: Any] = [:]
    metadata["services"] = config.services.map { $0.rawValue }

    if isServiceEnabled(.analytics) {
      var analytics: [String: Any] = [:]
      if let event {
        analytics["event"] = ["name": event.name, "properties": event.properties]
      }
      if !userProperties.isEmpty {
        analytics["userProperties"] = userProperties
      }
      let session = sessionPayload()
      if !session.isEmpty {
        analytics["session"] = session
      }
      if !analytics.isEmpty {
        metadata["analytics"] = analytics
      }
    }

    if isServiceEnabled(.errors), let errorInfo {
      var errorPayload: [String: Any] = ["message": errorInfo.message]
      if let stackTrace = errorInfo.stackTrace {
        errorPayload["stackTrace"] = stackTrace
      }
      if !errorInfo.properties.isEmpty {
        errorPayload["properties"] = errorInfo.properties
      }
      metadata["errors"] = errorPayload
    }

    if isServiceEnabled(.distribution) {
      let distribution = distributionPayload()
      if !distribution.isEmpty {
        metadata["distribution"] = distribution
      }
    }

    if isServiceEnabled(.devices) {
      let deviceInfo = devicePayload()
      if !deviceInfo.isEmpty {
        metadata["device"] = deviceInfo
      }
    }

    if isServiceEnabled(.usage) {
      metadata["usage"] = usagePayload()
    }

    if isServiceEnabled(.crashes), let crash = crash {
      var crashPayload = crash
      if crashPayload["appVersion"] == nil, let version = appVersion() {
        crashPayload["appVersion"] = version
      }
      if crashPayload["buildNumber"] == nil, let build = buildNumber() {
        crashPayload["buildNumber"] = build
      }
      if crashPayload["platform"] == nil {
        crashPayload["platform"] = config.platform
      }
      metadata["crash"] = crashPayload
    }

    if let custom = config.metadata, !custom.isEmpty {
      metadata["custom"] = custom
    }

    if !metadata.isEmpty {
      payload["metadata"] = metadata
    }

    guard JSONSerialization.isValidJSONObject(payload) else {
      throw EinstoreTrackingError.invalidPayload
    }
    return payload
  }

  private func sessionPayload() -> [String: Any] {
    let durationMs = Int(Date().timeIntervalSince(sessionStart) * 1000)
    return [
      "id": sessionId,
      "startedAt": isoString(from: sessionStart),
      "durationMs": durationMs,
    ]
  }

  private func usagePayload() -> [String: Any] {
    let now = Date()
    let offsetMinutes = TimeZone.current.secondsFromGMT(for: now) / 60
    return [
      "timestamp": isoString(from: now),
      "timeZone": TimeZone.current.identifier,
      "timeZoneOffsetMinutes": offsetMinutes,
      "locale": Locale.current.identifier,
      "sessionId": sessionId,
      "sessionDurationMs": Int(now.timeIntervalSince(sessionStart) * 1000),
    ]
  }

  private func distributionPayload() -> [String: Any] {
    var payload = config.distributionInfo
    mergeIfMissing(key: "appVersion", into: &payload, value: appVersion())
    mergeIfMissing(key: "buildNumber", into: &payload, value: buildNumber())
    return payload
  }

  private func devicePayload() -> [String: Any] {
    var payload = config.deviceInfo
    #if canImport(UIKit)
    mergeIfMissing(key: "model", into: &payload, value: UIDevice.current.model)
    mergeIfMissing(key: "manufacturer", into: &payload, value: "Apple")
    mergeIfMissing(key: "osVersion", into: &payload, value: UIDevice.current.systemVersion)
    #endif
    mergeIfMissing(key: "locale", into: &payload, value: Locale.current.identifier)
    mergeIfMissing(key: "appVersion", into: &payload, value: appVersion())
    mergeIfMissing(key: "buildNumber", into: &payload, value: buildNumber())
    return payload
  }

  private func appVersion() -> String? {
    Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String
  }

  private func buildNumber() -> String? {
    Bundle.main.infoDictionary?["CFBundleVersion"] as? String
  }

  private func mergeIfMissing(key: String, into payload: inout [String: Any], value: String?) {
    guard payload[key] == nil, let value else {
      return
    }
    payload[key] = value
  }

  private func isServiceEnabled(_ service: EinstoreService) -> Bool {
    config.services.contains(service)
  }

  private func isoString(from date: Date) -> String {
    ISO8601DateFormatter().string(from: date)
  }

  private func send(
    url: URL,
    payload: [String: Any],
    completion: ((Result<Void, Error>) -> Void)?
  ) {
    var request = URLRequest(url: url)
    request.httpMethod = "POST"
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    for (key, value) in config.headers {
      request.setValue(value, forHTTPHeaderField: key)
    }

    do {
      request.httpBody = try JSONSerialization.data(withJSONObject: payload, options: [])
    } catch {
      completion?(.failure(error))
      return
    }

    session.dataTask(with: request) { _, response, error in
      if let error {
        completion?(.failure(error))
        return
      }
      guard let http = response as? HTTPURLResponse else {
        completion?(.failure(EinstoreTrackingError.invalidResponse))
        return
      }
      guard (200...299).contains(http.statusCode) else {
        completion?(.failure(EinstoreTrackingError.requestFailed(statusCode: http.statusCode)))
        return
      }
      completion?(.success(()))
    }.resume()
  }
}
